src/core/lib/transport/metadata_batch.h

```diff
@@ -185,6 +190,88 @@ class MetadataMap {
       GRPC_MUST_USE_RESULT;
   void MaybeUnlinkCallout(grpc_linked_mdelem* storage);
 
+  static void assert_valid_list(grpc_mdelem_list* list) {
+#ifndef NDEBUG
+    grpc_linked_mdelem* l;
+
+    GPR_ASSERT((list->head == nullptr) == (list->tail == nullptr));
+    if (!list->head) return;
+    GPR_ASSERT(list->head->prev == nullptr);
+    GPR_ASSERT(list->tail->next == nullptr);
+    GPR_ASSERT((list->head == list->tail) == (list->head->next == nullptr));
+
+    size_t verified_count = 0;
+    for (l = list->head; l; l = l->next) {
+      GPR_ASSERT(!GRPC_MDISNULL(l->md));
+      GPR_ASSERT((l->prev == nullptr) == (l == list->head));
+      GPR_ASSERT((l->next == nullptr) == (l == list->tail));
+      if (l->next) GPR_ASSERT(l->next->prev == l);
+      if (l->prev) GPR_ASSERT(l->prev->next == l);
+      verified_count++;
+    }
+    GPR_ASSERT(list->count == verified_count);
+#else
+    // Avoid unused-parameter warning for debug-only parameter
+    (void)list;
+#endif /* NDEBUG */
+  }
+
+  static grpc_error_handle GPR_ATTRIBUTE_NOINLINE
+  error_with_md(grpc_mdelem md) {
+    return grpc_attach_md_to_error(
+        GRPC_ERROR_CREATE_FROM_STATIC_STRING("Unallowed duplicate metadata"),
+        md);
+  }
+
+  static void link_head(grpc_mdelem_list* list, grpc_linked_mdelem* storage) {
+    assert_valid_list(list);
+    GPR_DEBUG_ASSERT(!GRPC_MDISNULL(storage->md));
+    storage->prev = nullptr;
+    storage->next = list->head;
+    storage->reserved = nullptr;
+    if (list->head != nullptr) {
+      list->head->prev = storage;
+    } else {
+      list->tail = storage;
+    }
+    list->head = storage;
+    list->count++;
+    assert_valid_list(list);
```

Comment:
I'm curious in why this second valid check is necessary? How can a list go from valid -> invalid in this function? Same question in below functions

Reply:
I remember having some difficulty getting this linked list code right, and so added these whenever we manipulated the list, as a pre and post condition. That way, if I got it wrong, I'd have a boundary back to where it was last correct to focus my bug search.

Reply:
Hm if this code is sticking around, it would be useful to have a comment here. If not, ignore me :)

Reply:
It'll be gone soon... towards the end of this transition, but definitely scoped inside of it. Plan is to replace this linked list with something like `InlinedVector<pair<Slice, Slice>>`

Reply:
Ok sg, thanks!